[rpc] modify test_local_shutdown_with_rpc to not be flaky

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30640 [rpc] modify test_local_shutdown_with_rpc to not be flaky**

This test would get very occasional flakes, with an error saying the
RPC timed out. This happened because one worker would still be waiting for the
return value of an RPC, but another worker had already performed its local
shutdown, so it would not have sent the response. This didn't show up in
initial testing since the flakiness is very rare (< 1/100 test runs). We can add a dist.barrier() that all processes call into to alleviate the issue.

Differential Revision: [D18775731](https://our.internmc.facebook.com/intern/diff/D18775731/)

rohan-varma committed:
[rpc] modify test_local_shutdown_with_rpc to not be flaky

This test would get very occasional flakes, with an error saying the
RPC timed out. This happened because one worker would still be waiting for the
return value of an RPC, but another worker had already performed its local
shutdown, so it would not have sent the response. This didn't show up in
initial testing since the flakiness is very rare (< 1/100 test runs). This diff
fixes the issue by not erroring if these RPCs timeout. The reason this is okay
is because with a local shutdown, we should not expect for all outstanding RPCs
to be completed, since workers are free to shut down without completing/waiting
on outstanding work.

Differential Revision: [D18775731](https://our.internmc.facebook.com/intern/diff/D18775731/)

[ghstack-poisoned]

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -1193,17 +1177,20 @@ def test_local_shutdown_with_rpc(self):
+            )
+        except RuntimeError as e:
+            self.assertTrue("RPC ran for more than" in str(e))
         # pass in graceful=False to ensure that we don't wait for other workers.
```
Can't we just add a `dist.barrier` before the shutdown to fix the problem?

rohan-varma replied:
Good idea, I think this is a better solution.

rohan-varma committed:
Update on "[rpc] modify test_local_shutdown_with_rpc to not be flaky"


This test would get very occasional flakes, with an error saying the
RPC timed out. This happened because one worker would still be waiting for the
return value of an RPC, but another worker had already performed its local
shutdown, so it would not have sent the response. This didn't show up in
initial testing since the flakiness is very rare (< 1/100 test runs). This diff
fixes the issue by not erroring if these RPCs timeout. The reason this is okay
is because with a local shutdown, we should not expect for all outstanding RPCs
to be completed, since workers are free to shut down without completing/waiting
on outstanding work.

Differential Revision: [D18775731](https://our.internmc.facebook.com/intern/diff/D18775731/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] modify test_local_shutdown_with_rpc to not be flaky"


This test would get very occasional flakes, with an error saying the
RPC timed out. This happened because one worker would still be waiting for the
return value of an RPC, but another worker had already performed its local
shutdown, so it would not have sent the response. This didn't show up in
initial testing since the flakiness is very rare (< 1/100 test runs). We can add a dist.barrier() that all processes call into to alleviate the issue.

Differential Revision: [D18775731](https://our.internmc.facebook.com/intern/diff/D18775731/)

[ghstack-poisoned]

xush6528 commented on test/rpc_test.py:
```diff
@@ -1132,15 +1116,22 @@ def test_local_shutdown_with_rpc(self):
+                rank=self.rank,
+                world_size=self.world_size,
+            )
+        dist.barrier()
```
I think it's easier to understand than an ad-hoc helper function, `_check_rpc_done`, put in this file.

Still not too perfect to initialize process group. I think later, if `_wai_all_workers()` supports calling multiple times. We could use it.

rohan-varma committed:
Update on "[rpc] modify test_local_shutdown_with_rpc to not be flaky"


This test would get very occasional flakes, with an error saying the
RPC timed out. This happened because one worker would still be waiting for the
return value of an RPC, but another worker had already performed its local
shutdown, so it would not have sent the response. This didn't show up in
initial testing since the flakiness is very rare (< 1/100 test runs). We can add a dist.barrier() that all processes call into to alleviate the issue.

Differential Revision: [D18775731](https://our.internmc.facebook.com/intern/diff/D18775731/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] modify test_local_shutdown_with_rpc to not be flaky"


This test would get very occasional flakes, with an error saying the
RPC timed out. This happened because one worker would still be waiting for the
return value of an RPC, but another worker had already performed its local
shutdown, so it would not have sent the response. This didn't show up in
initial testing since the flakiness is very rare (< 1/100 test runs). We can add a dist.barrier() that all processes call into to alleviate the issue.

Differential Revision: [D18775731](https://our.internmc.facebook.com/intern/diff/D18775731/)

[ghstack-poisoned]

rohan-varma commented:
Closing in favor of https://github.com/pytorch/pytorch/pull/30837, since this is having rebase errors according to github.
